src/hardware/agilent-54621d/protocol.c

```diff
@@ -0,0 +1,1032 @@
+	char *tmp_str;
+
+	devc = sdi->priv;
+	config = devc->model_config;
+	state = devc->model_state;
+
+	sr_info("Fetching scope state");
+
+	//get analog channel state
+	if (analog_channel_state_get(sdi, config, state) != SR_OK)
+		return SR_ERR;
+
+	//get digital channel state
+	if (digital_channel_state_get(sdi, config, state) != SR_OK)
+		return SR_ERR;
+
+	//get timebase
+	if (sr_scpi_get_string(sdi->conn,
+			(*config->scpi_dialect)[SCPI_CMD_GET_TIMEBASE],
+			&tmp_str) != SR_OK)
+		return SR_ERR;
+
+	if (array_float_get(tmp_str, ARRAY_AND_SIZE(scope_timebases), &i) != SR_OK) {
+		g_free(tmp_str);
+		sr_err("Could not determine array index for time base.");
+		return SR_ERR;
+	}
+	g_free(tmp_str);
+
+	state->timebase = i;
+
+	//get trigger horizontal position
+	if(sr_scpi_get_float(sdi->conn, (*config->scpi_dialect)[SCPI_CMD_GET_HORIZ_TRIGGERPOS], &tmp_float) != SR_OK)
+		return SR_ERR;
+	
+	state->horiz_triggerpos = tmp_float /
+		(((double) (*config->timebases)[state->timebase][0] /
+		  (*config->timebases)[state->timebase][1]) * config->num_xdivs);
+	state->horiz_triggerpos -= 0.5;
+	state->horiz_triggerpos *= -1;
+	//ToDo: This might be dependent on time ref setting
+
+	if (scope_state_get_array_option(sdi->conn,	(*config->scpi_dialect)[SCPI_CMD_GET_TRIGGER_SOURCE], config->trigger_sources, config->num_trigger_sources, &state->trigger_source) != SR_OK)
+		return SR_ERR;
+
+	if (scope_state_get_array_option(sdi->conn, (*config->scpi_dialect)[SCPI_CMD_GET_TRIGGER_SLOPE], config->trigger_slopes, config->num_trigger_slopes, &state->trigger_slope) != SR_OK)
+		return SR_ERR;
+
+	//ToDo: get trigger pattern
+	//documentation for reading the trigger pattern is a little wonky so I need to test how this is done
+	strncpy(state->trigger_pattern, "00000000000000000000", MAX_ANALOG_CHANNEL_COUNT + MAX_DIGITAL_CHANNEL_COUNT+1); 
+
+
+	//ToDo: get current resolution
+	//Default resolution is 8bit. aquiring at 8 bit also can increase transfer speed, since only a byte of data per point has to be transmitted
+	//Resolution is a function of sweep speed and number of averages
+	//Resolution > 8bit if acq mode == avg && (timebase >= 5us/div || num_avg>1) 
+	state->high_resolution = FALSE;
+
+
+	//get peak detection
+	if (sr_scpi_get_string(sdi->conn,
+			     (*config->scpi_dialect)[SCPI_CMD_GET_PEAK_DETECTION],
+			     &tmp_str) != SR_OK)
+		return SR_ERR;
+	if (!strcmp("PEAK", tmp_str))
+		state->peak_detection = TRUE;
+	else
+		state->peak_detection = FALSE;
+	g_free(tmp_str);
+
+	if(agilent_54621d_update_sample_rate(sdi) != SR_OK)
+		return SR_ERR;
+
+	sr_info("Fetching finished.");
+
+	scope_state_dump(config, state);
+
+	return SR_OK;
+}
+
+static int scope_state_get_array_option(struct sr_scpi_dev_inst *scpi, const char *command, const char *(*array)[], unsigned int n, int *result)
+{
+	char *tmp;
+	int idx;
+
+	if (sr_scpi_get_string(scpi, command, &tmp) != SR_OK)
+		return SR_ERR;
+
+	if ((idx = std_str_idx_s(tmp, *array, n)) < 0) {
```

Comment:
splitting assignment out of conditional would also allow you to `g_free(tmp)` immediately after, in one place only